### ISSUE TITLE
How to deal with Unicode filenames?

### ISSUE THREAD

**Issue by joepie91** [jjjake/internetarchive#63]
Upon encountering a filename that contains Unicode(?) characters, the following occurs:

```
Traceback (most recent call last):
  File "./mirror.py", line 67, in <module>
    if item.upload([(real_filename, source_file)], metadata=metadata, access_key=conf["internetarchive"]["accesskey"], secret_key=conf["internetarchive"]["secretkey"]):
  File "/var/sites/pdfhost/internetarchive/item.py", line 544, in upload
    resp = self.upload_file(body, key=key, **kwargs)
  File "/var/sites/pdfhost/internetarchive/item.py", line 419, in upload_file
    url = '{base_url}/{key}'.format(base_url=base_url, key=key)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 13-14: ordinal not in range(128)
```

Evidently, it doesn't like these characters. What is the correct way of dealing with this, that doesn't upset either the library or the IA HTTPd? Is it possible at all to have non-ASCII characters in filenames on IA?

For reference, [this](http://pdf.yt/d/LxGP4MUquejAAo0N) is the offending document (see the filename in the sidebar).

**Comment by CarlFK** [comment]
I have been bumping into unicode issues.
I forked, so yes, this is a tainted code, but after an hour or so of struggling it would be nice if someone else could take a look and tell me what's going on.

```
$ python 
Python 2.7.9 (default, Apr  2 2015, 15:33:21) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from internetarchive import get_item
>>> i = get_item('iacli-test-item9')
>>> i
Item(identifier='iacli-test-item9', exists=False)
>>> fn = u'/home/carl/Videos/veyepar/pyconza/pyconza2015/mp4/PyCon_Montréal_in_30_min.mp4'
>>> fn
u'/home/carl/Videos/veyepar/pyconza/pyconza2015/mp4/PyCon_Montr\xe9al_in_30_min.mp4'
>>> i.upload(fn)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/carl/.virtualenvs/veyepar/local/lib/python2.7/site-packages/internetarchive/item.py", line 651, in upload
    resp = self.upload_file(body, key=key, **kwargs)
  File "/home/carl/.virtualenvs/veyepar/local/lib/python2.7/site-packages/internetarchive/item.py", line 575, in upload_file
    raise type(exc)(error_msg)
requests.exceptions.HTTPError
>>> from internetarchive import __version__; print(__version__)
0.9.6
>>> 
```

**Comment by CarlFK** [comment]
Hmm.. tried it on a clean box and don't have creds:
```
>>> i.upload(fn)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "internetarchive/item.py", line 634, in upload
    resp = self.upload_file(body, key=key, **kwargs)
  File "internetarchive/item.py", line 558, in upload_file
    raise type(exc)(error_msg)
requests.exceptions.HTTPError:  error uploading PyCon_Montréal_in_30_min.txt to iacli-test-item9, 403 Client Error: Forbidden
```

What I don't understand is why I don't get that on the first box because I never did the CLI thing to setup creds.  grumble.  sorry for dumping crap in the issue tracker, I wanted a cleaner report but ran out of time.

**Comment by CarlFK** [comment]
I think there are two things that need to handle unicode:  Identifier and file name.
I think my branch handles file name.  But when I use an archive identifier with a unicode char I get:
```
ERROR:internetarchive.item: error uploading PyCon_Montr?al_in_30_min.mp4 to PyCon_Montréal, 400 Client Error: Bad Request
```
I don't have a simple >>> steps yet.  I have been using my video uploader driver as a starting point.  
and now I need to get some sleep, so this may have to wait for more debugging.

**Comment by CarlFK** [comment]
Here is a simple example of an error when  the identifier has unicode chars:

```
>>> from pw import archive
>>> auth = archive['test']
>>> 
>>> from internetarchive import get_item
>>> i = get_item(u'Montréal_1')
>>> i
Item(identifier=u'Montr\xe9al_1', exists=False)
>>> i.upload(u'/tmp/pycon_1.txt', access_key=auth['access'], secret_key=auth['secret'], verbose=True )
Python 2.7.9 (default, Apr  2 2015, 15:33:21) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> exc
HTTPError('400 Client Error: Bad Request',)
>>> 
```
This is perhaps a separate issue from filename, let me know if you want me to open a new github issue.

ps - the example ia pages no longer work:
https://archive.org/download/iacli-test-item3/
"The item is not available due to issues with the item's content."

**Comment by CarlFK** [comment]
http://archive.org/about/faqs.php#1099             
>How should I name my files?
>Support for Filenames is limited to basic ASCII characters.
>Please use only unaccented letters, numbers, dashes, underscores or periods.
>Do not use special characters 

I think it is reasonable to assume this applies to identifiers too.

**Comment by Asparagirl** [comment]
I just ran into this one too while trying to upload in bulk through a spreadsheet (via the ```ia upload --spreadsheet=uploading.csv``` method).

I know this issue is marked "won't fix", but could we maybe get a simpler-to-read error message in cases like this, just something like "It looks like you're trying to do a bulk upload with a spreadsheet, please make sure to convert your CSV file to ASCII format first, not to UTF-8"?

**Comment by saper** [comment]
@Asparagirl did you use utf-8 in the description and other metadata or the file name?

**Comment by Asparagirl** [comment]
The weird thing is that I _wasn't_ using any unicode or accented characters in my spreadsheet, as far as I know.  I did have some HTML in my description column, but it was all just plain English and some basic HTML code.

I first saved the spreadsheet as a CSV with normal formatting, which in NeoOffice means UTF-8, and that was throwing the errors.  Saving the spreadsheet as a Western/ASCII-formatted CSV fixed the problem. 

Basically, it would be nice for the code to either handle UTF-8 formatted spreadsheets, or else just give a more easily understood error message, letting the user know to save the CSV as ASCII.  And maybe a line or two about this in the docs somewhere.  That's all.

**Comment by jjjake** [comment]
@Asparagirl @saper `internetarchive` has full support for uploading utf-8 files and adding utf-8 metadata values as far as I can tell. I'm not an expert in this area, so I would love to be corrected if I'm misunderstanding anything here.

I just created the following items with `ia upload --spreadsheet metadata.csv`:

https://archive.org/details/jj-test-item-2017-01-06-3
https://archive.org/details/jj-test-item-2017-01-06-4

The following files were uploaded with `--spreadsheet`:

https://archive.org/download/jj-test-item-2017-01-06-3/%e0%ae%99%e0%af%8d-%d0%9b-%d0%81-%e3%81%8e-%e3%81%8e_utf-8.txt
https://archive.org/download/jj-test-item-2017-01-06-4/%e0%ae%99%e0%af%8d-%d0%9b-%d0%81-%e3%81%8e-%e3%81%8e_utf-8.txt

This file was uploaded with just `ia upload <id> <file>`:

https://archive.org/download/jj-test-item-2017-01-06-3/%e0%ae%99%e0%af%8d-%d0%9b-%d0%95%cc%88-%e3%81%8d%e3%82%99-%e3%81%8d%e3%82%99-2_utf-8.txt

You'll see that each item has various utf-8 encoded special characters.

The spreadsheet I used to create these items is here:

https://archive.org/download/jj-test-item-2017-01-06-3/metadata.csv

This was all done with the latest version of `internetarchive`, v1.1.0, on Python 2.7.12 on Mac OS X 10.11.4. The spreadsheet is utf-8 encoded and was generated with libre office.

I'm not sure if this would matter, but my locale settings look like this:

    » locale
    LANG="en_US.UTF-8"
    LC_COLLATE="en_US.UTF-8"
    LC_CTYPE="en_US.UTF-8"
    LC_MESSAGES="en_US.UTF-8"
    LC_MONETARY="en_US.UTF-8"
    LC_NUMERIC="en_US.UTF-8"
    LC_TIME="en_US.UTF-8"
    LC_ALL=

I don't mean to write off your issues, so my apologies if I come off that way! : )

I'm just having trouble replicating these issues myself. I'll try to replicate these issues in different environments when I get a chance.

If anyone is able to provide a copy of the spreadsheets or files they're having issues with, that would be wonderful.

Thanks again for taking the time to detail your issues and report them here @saper and @Asparagirl. I hope we can get to the bottom of this.

**Comment by Asparagirl** [comment]
Hello again,

I uploaded my original .ODS spreadsheet and the two versions of the exported .CSV's (one UTF-8 and one ASCII) to Dropbox, so feel free to grab them and look for yourself:

https://www.dropbox.com/sh/zm9801360niad00/AAAbWYforhGFKKQp_Knhy-Oza?dl=0

As you can see, ```ia upload``` chokes on the UTF-8 one.  It does parse and upload the ASCII one...well, until I get the "broken pipe" issue as described in #176.

My locale settings match yours.

Good luck, soldier.  :-)

**Comment by jjjake** [comment]
@Asparagirl How does `ia upload` choke on the UTF-8 spreadsheet for you? I changed the collection and the identifier on the spreadsheet, and created a few ZIPs with the same filenames as you had on the spreadsheet. It uploaded without issue. What kind of errors are you seeing with the UTF-8 spreadsheet? Are you sure it's unicode related? The ASCII spreadsheet in your dropbox link is a 404 for me, by the way.

**Comment by TheCoinCollector** [comment]
Hey, I was having this same problem on the current version, internetarchive 1.5.0 - in my case it was a TM symbol tripping it up.  I was able to fix it for me with @saper 's fix plus one other similar change.

My original error:
```

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/bin/ia", line 11, in <module>
    sys.exit(main())
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/internetarchive/cli/ia.py", line 159, in main
    sys.exit(ia_module.main(argv, session))
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/internetarchive/cli/ia_upload.py", line 234, in main
    md_args = ['{0}:{1}'.format(k.lower(), v) for (k, v) in row.items() if v]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 24: ordinal not in range(128)
```
I did @saper 's change, in ia_upload.py:
`md_args = ['{0}:{1}'.format(k.lower(), v) for (k, v) in row.items() if v]`
to
`md_args = ['{0}:{1}'.format(k.decode("utf-8").lower(), v.decode("utf-8")) for (k, v) in row.items() if v]`

Then I got this error:
```

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/bin/ia", line 11, in <module>
    sys.exit(main())
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/internetarchive/cli/ia.py", line 159, in main
    sys.exit(ia_module.main(argv, session))
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/internetarchive/cli/ia_upload.py", line 238, in main
    session)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/internetarchive/cli/ia_upload.py", line 78, in _upload_files
    response = item.upload(files, **upload_kwargs)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/internetarchive/item.py", line 771, in upload
    request_kwargs=request_kwargs)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/internetarchive/item.py", line 549, in upload_file
    key = norm_filepath(filename).split('/')[-1] if key is None else key
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 25: ordinal not in range(128)
```

And in item.py I changed:
`key = norm_filepath(filename).split('/')[-1] if key is None else key`
to
`key = norm_filepath(filename).decode("utf-8").split('/')[-1] if key is None else key`

I've never used GitHub before so apologies if I should be conveying this with a fork comparison or whatever it would be!  I wonder why some of us are getting this error and others are fine.  I'm on Mac OS X 10.11.6.

**Comment by jjjake** [comment]
I think a586b96 should resolve this issue.

Please let me know if this does not resolve the issue for you, and I'll reopen the issue. Thanks again, and sorry for the **very** long wait! : )

**Comment by Asparagirl** [comment]
I don't have any big files to upload right now, but I'll let you know how it goes the next time I do.  😄

**Comment by danaronson** [comment]
this still seems to be a problem.  just trying to retrieve an item using 1.7.2dev1

```
Justins-Air:internetarchive dan$ ia --version
1.7.2.dev1
Justins-Air:internetarchive dan$ python
Python 2.7.12 |Anaconda 4.2.0 (x86_64)| (default, Jul  2 2016, 17:43:17)
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> from internetarchive import get_item
>>> get_item('cd_qunxing‑sh‑yangzong‑linzhi‑shabaoliang‑yuq')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "internetarchive/api.py", line 116, in get_item
    return archive_session.get_item(identifier, request_kwargs=request_kwargs)
  File "internetarchive/session.py", line 213, in get_item
    'retrieving now.'.format(identifier))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 10: ordinal not in range(128)
```

**Comment by jjjake** [comment]
@danaronson Your identifier is not ascii. All identifiers should be ascii. It looks like your hyphens are actually non-breaking hyphens.

Try this:

    >>> from internetarchive import get_item
    >>> get_item('cd_qunxing-sh-yangzong-linzhi-shabaoliang-yuq')
    Item(identifier='cd_qunxing-sh-yangzong-linzhi-shabaoliang-yuq')

I understand how this might have been confusing. Please feel free to create a new issue with any suggestions you might have for improving the way issues like these are handled. Thanks for reporting, I appreciate it!

**Comment by danaronson** [comment]
you are right indeed.  sorry for the false alarm.  I cut and pasted for a web display of the identifier, not quite sure why that display was that.
